File: web/src/lib/analysisEngine.js
// Centralized financial analysis engine. Single source of truth for
// the dashboard's eight summary metrics — current values, prior-
// window values, and percent deltas — so the same numbers can be
// rendered by web's DashboardClient and the mobile app (via the
// /api/analysis HTTP endpoint).
//
// Inputs:
//   receipts      — public.receipts rows
//   statements    — public.bank_statements rows (or [])
//   fees          — public.bank_fees rows (or [])
//   transactions  — public.bank_transactions rows (or [])
//   period        — 'daily' | 'weekly' | 'monthly' | 'yearly'
//   periodCount   — number of period units (e.g. 3 = "last 3 months")
//
// Output:
//   {
//     period, periodCount,
//     rangeStart, rangeEnd,                  (YYYY-MM-DD strings)
//     metrics: {
//       transactions, totalSpent, taxPaid, bankFees,    (receipt-side)
//       purchases, payments, interestPaid, feesPaid,    (bank-side)
//     }
//   }
// Each metric: { current, prior, deltaPct, deltaArrow, deltaLabel }
//   deltaPct    — number | null (null when prior is 0, can't divide)
//   deltaArrow  — '↑' | '↓' | '→' | null
//   deltaLabel  — short human badge like '↑ 12%' or '—' for zero prior

import { subDays, subWeeks, subMonths, subYears } from 'date-fns'
import { isPaymentReceipt } from './payment-rows'

// ── Period helpers ──────────────────────────────────────────────────
function periodStart(period, count) {
  const now = new Date()
  if (period === 'daily')   return subDays(now,  count)
  if (period === 'weekly')  return subWeeks(now, count)
  if (period === 'monthly') return subMonths(now, count)
  if (period === 'yearly')  return subYears(now, count)
  return now
}

function toIsoDate(d) {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

// ── Delta calculation ───────────────────────────────────────────────
function buildDelta(current, prior) {
  if (!prior) {
    return { deltaPct: null, deltaArrow: null, deltaLabel: '—' }
  }
  const pct = ((current - prior) / prior) * 100
  const arrow = pct > 1 ? '↑' : pct < -1 ? '↓' : '→'
  const label = arrow === '→' ? '—' : `${arrow} ${Math.abs(pct).toFixed(0)}%`
  return { deltaPct: pct, deltaArrow: arrow, deltaLabel: label }
}

function metric(current, prior) {
  return { current, prior, ...buildDelta(current, prior) }
}

// ── Range filters ───────────────────────────────────────────────────
function filterByDateStr(rows, dateField, startStr, endStr) {
  return rows.filter(r => {
    const d = String(r[dateField] || '')
    if (d.length < 10) return false
    if (startStr && d < startStr) return false
    if (endStr   && d >= endStr)  return false
    return true
  })
}

// ── Bank-side aggregation for a single window ────────────────────────
// Mirrors the categorisation logic in financeInsights.bankAccountTotals
// but works on a date-string range so we can compute current and prior
// windows in one place without touching that function's period-key API.
function bankWindowTotals({ fees, transactions }, startStr, endStr) {
  const feeRows = filterByDateStr(fees, 'date', startStr, endStr)
  const txnRows = filterByDateStr(transactions, 'date', startStr, endStr)

  const sumAbs = (rows) => rows.reduce((n, x) => n + Math.abs(Number(x.amount || 0)), 0)
  const sumPos = (rows) => rows.reduce((n, x) => n + Number(x.amount || 0), 0)

  const paymentRows = txnRows.filter(t => t.is_payment)

  return {
    interest: sumAbs(feeRows.filter(f => f.kind === 'interest')) + sumAbs(txnRows.filter(t => t.is_interest)),
    fees:     sumAbs(feeRows.filter(f => f.kind === 'fee' || f.kind === 'penalty')) + sumAbs(txnRows.filter(t => t.is_fee)),
    payments: sumAbs(paymentRows),
    purchases: sumPos(txnRows.filter(t => !t.is_payment && !t.is_fee && !t.is_interest && !t.is_refund && t.amount > 0)),
  }
}

// ── Receipt-side window totals ──────────────────────────────────────
function receiptWindowTotals(receipts, startStr, endStr) {
  const rows = filterByDateStr(receipts, 'date', startStr, endStr)
  return {
    transactions: rows.length,
    totalSpent:   rows.reduce((s, r) => s + parseFloat(r.total_amount || 0), 0),
    taxPaid:      rows.reduce((s, r) => s + parseFloat(r.tax_paid || 0),     0),
    bankFees:     rows
                    .filter(r => r.category === 'bank-fees')
                    .reduce((s, r) => s + parseFloat(r.total_amount || 0), 0),
  }
}

// ── Public API ──────────────────────────────────────────────────────
export function computeDashboardAnalysis({
  receipts = [],
  statements = [],
  fees = [],
  transactions = [],
  period = 'monthly',
  periodCount = 3,
} = {}) {
  // Drop card-payment + transfer receipts before any spending math —
  // same filter the dashboard already applies. Mirroring it here means
  // mobile gets the same numbers without re-implementing the rule.
  const spendingReceipts = receipts.filter(r => !isPaymentReceipt(r))

  const currentStart = toIsoDate(periodStart(period, periodCount))
  const priorStart   = toIsoDate(periodStart(period, periodCount * 2))
  // No upper bound for the current window — anything from currentStart
  // forward (including today) counts. The prior window ends where the
  // current one starts.

  const curr = receiptWindowTotals(spendingReceipts, currentStart, null)
  const prev = receiptWindowTotals(spendingReceipts, priorStart, currentStart)

  const currBank = bankWindowTotals({ fees, transactions }, currentStart, null)
  const prevBank = bankWindowTotals({ fees, transactions }, priorStart, currentStart)

  return {
    period,
    periodCount,
    rangeStart: currentStart,
    rangeEnd:   null,
    priorRangeStart: priorStart,
    priorRangeEnd:   currentStart,
    metrics: {
      transactions: metric(curr.transactions, prev.transactions),
      totalSpent:   metric(curr.totalSpent,   prev.totalSpent),
      taxPaid:      metric(curr.taxPaid,      prev.taxPaid),
      bankFees:     metric(curr.bankFees,     prev.bankFees),
      purchases:    metric(currBank.purchases, prevBank.purchases),
      payments:     metric(currBank.payments,  prevBank.payments),
      interestPaid: metric(currBank.interest,  prevBank.interest),
      feesPaid:     metric(currBank.fees,      prevBank.fees),
    },
  }
}

File: mobile/lib/services/analysis_engine.dart
// Centralized dashboard analysis — Dart port of
// web/src/lib/analysisEngine.js. Single source of truth for the
// mobile dashboard's eight financial metrics, current vs. prior
// period delta included, so the iOS/Android app and the web app
// show the same numbers for the same window.
//
// Two entry points:
//   - computeDashboardAnalysis(...) — pure function over already-
//     loaded data. Use this if the caller already has receipts +
//     bank rows in memory.
//   - fetchDashboardAnalysis(...)   — convenience wrapper that
//     pulls statements / fees / transactions from Supabase (RLS-
//     scoped) and runs the engine.

import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/receipt_model.dart';
import '../payment_rows.dart';

class MetricValue {
  final double current;
  final double prior;
  final double? deltaPct;
  final String? deltaArrow;
  final String deltaLabel;
  const MetricValue({
    required this.current,
    required this.prior,
    this.deltaPct,
    this.deltaArrow,
    required this.deltaLabel,
  });
}

class DashboardAnalysis {
  final String period;
  final int periodCount;
  final String rangeStart;
  final String priorRangeStart;
  final Map<String, MetricValue> metrics;
  const DashboardAnalysis({
    required this.period,
    required this.periodCount,
    required this.rangeStart,
    required this.priorRangeStart,
    required this.metrics,
  });
}

class BankData {
  final List<Map<String, dynamic>> statements;
  final List<Map<String, dynamic>> fees;
  final List<Map<String, dynamic>> transactions;
  const BankData({
    required this.statements,
    required this.fees,
    required this.transactions,
  });
  static const empty = BankData(statements: [], fees: [], transactions: []);
}

// ── Period helpers ─────────────────────────────────────────────────
DateTime periodStartDate(String period, int count) {
  final now = DateTime.now();
  final n = count < 1 ? 1 : count;
  switch (period) {
    case 'daily':   return DateTime(now.year, now.month, now.day - n);
    case 'weekly':  return DateTime(now.year, now.month, now.day - (n * 7));
    case 'monthly': return DateTime(now.year, now.month - n, now.day);
    case 'yearly':  return DateTime(now.year - n, now.month, now.day);
    default:        return now;
  }
}

String toIsoDate(DateTime d) {
  final m  = d.month.toString().padLeft(2, '0');
  final dd = d.day.toString().padLeft(2, '0');
  return '${d.year}-$m-$dd';
}

String periodStartIso(String period, int count) =>
    toIsoDate(periodStartDate(period, count));

// ── Delta calculation — mirrors web's buildDelta ─────────────────────
MetricValue _metric(double current, double prior) {
  if (prior == 0) {
    return MetricValue(current: current, prior: prior, deltaLabel: '—');
  }
  final pct = ((current - prior) / prior) * 100;
  final arrow = pct > 1 ? '↑' : pct < -1 ? '↓' : '→';
  final label = arrow == '→' ? '—' : '$arrow ${pct.abs().toStringAsFixed(0)}%';
  return MetricValue(
    current: current, prior: prior,
    deltaPct: pct, deltaArrow: arrow, deltaLabel: label,
  );
}

// ── Range-filter helpers ────────────────────────────────────────────
List<Map<String, dynamic>> _filterByDateStr(
  List<Map<String, dynamic>> rows,
  String dateField,
  String? startStr,
  String? endStr,
) {
  return rows.where((r) {
    final d = (r[dateField] ?? '').toString();
    if (d.length < 10) return false;
    if (startStr != null && d.compareTo(startStr) < 0) return false;
    if (endStr   != null && d.compareTo(endStr)   >= 0) return false;
    return true;
  }).toList();
}

List<Receipt> _filterReceiptsByDateStr(
  List<Receipt> rows,
  String? startStr,
  String? endStr,
) {
  return rows.where((r) {
    final d = r.date;
    if (d.length < 10) return false;
    if (startStr != null && d.compareTo(startStr) < 0) return false;
    if (endStr   != null && d.compareTo(endStr)   >= 0) return false;
    return true;
  }).toList();
}

// ── Receipt-side totals over a window ────────────────────────────────
class _ReceiptTotals {
  final int transactions;
  final double totalSpent;
  final double taxPaid;
  final double bankFees;
  const _ReceiptTotals({
    required this.transactions,
    required this.totalSpent,
    required this.taxPaid,
    required this.bankFees,
  });
}

_ReceiptTotals _receiptWindowTotals(List<Receipt> receipts, String? startStr, String? endStr) {
  final rows = _filterReceiptsByDateStr(receipts, startStr, endStr);
  double totalSpent = 0, taxPaid = 0, bankFees = 0;
  for (final r in rows) {
    totalSpent += r.totalAmount;
    taxPaid   += r.taxPaid;
    if (r.category == 'bank-fees') bankFees += r.totalAmount;
  }
  return _ReceiptTotals(
    transactions: rows.length,
    totalSpent: totalSpent,
    taxPaid: taxPaid,
    bankFees: bankFees,
  );
}

// ── Bank-side totals over a window — mirrors the categorisation in
//    web/src/lib/financeInsights.bankAccountTotals so the numbers
//    match what /api/analysis returns.
class _BankTotals {
  final double interest;
  final double fees;
  final double payments;
  final double purchases;
  const _BankTotals({
    required this.interest,
    required this.fees,
    required this.payments,
    required this.purchases,
  });
}

double _sumAbs(List<Map<String, dynamic>> rows, [bool Function(Map<String, dynamic>)? where]) {
  double s = 0;
  for (final r in rows) {
    if (where != null && !where(r)) continue;
    final v = double.tryParse((r['amount'] ?? 0).toString()) ?? 0;
    s += v.abs();
  }
  return s;
}

double _sumPos(List<Map<String, dynamic>> rows, [bool Function(Map<String, dynamic>)? where]) {
  double s = 0;
  for (final r in rows) {
    if (where != null && !where(r)) continue;
    final v = double.tryParse((r['amount'] ?? 0).toString()) ?? 0;
    s += v;
  }
  return s;
}

_BankTotals _bankWindowTotals(BankData bank, String? startStr, String? endStr) {
  final feeRows = _filterByDateStr(bank.fees, 'date', startStr, endStr);
  final txnRows = _filterByDateStr(bank.transactions, 'date', startStr, endStr);

  bool flag(Map<String, dynamic> m, String key) => m[key] == true;

  final interest = _sumAbs(feeRows, (f) => f['kind'] == 'interest')
                 + _sumAbs(txnRows, (t) => flag(t, 'is_interest'));
  final fees     = _sumAbs(feeRows, (f) => f['kind'] == 'fee' || f['kind'] == 'penalty')
                 + _sumAbs(txnRows, (t) => flag(t, 'is_fee'));
  final payments = _sumAbs(txnRows, (t) => flag(t, 'is_payment'));
  final purchases = _sumPos(txnRows, (t) {
    final amt = double.tryParse((t['amount'] ?? 0).toString()) ?? 0;
    return !flag(t, 'is_payment') && !flag(t, 'is_fee')
        && !flag(t, 'is_interest') && !flag(t, 'is_refund')
        && amt > 0;
  });
  return _BankTotals(interest: interest, fees: fees, payments: payments, purchases: purchases);
}

// ── Public API ──────────────────────────────────────────────────────
DashboardAnalysis computeDashboardAnalysis({
  required List<Receipt> receipts,
  required BankData bank,
  required String period,
  required int periodCount,
}) {
  // Same pre-filter the web engine applies — strip card-payment /
  // transfer rows so they don't get double-counted.
  final spendingReceipts = receipts.where((r) => !isPaymentReceipt(r)).toList();

  final currentStart = periodStartIso(period, periodCount);
  final priorStart   = periodStartIso(period, periodCount * 2);

  final curr = _receiptWindowTotals(spendingReceipts, currentStart, null);
  final prev = _receiptWindowTotals(spendingReceipts, priorStart, currentStart);
  final currBank = _bankWindowTotals(bank, currentStart, null);
  final prevBank = _bankWindowTotals(bank, priorStart, currentStart);

  return DashboardAnalysis(
    period: period,
    periodCount: periodCount,
    rangeStart: currentStart,
    priorRangeStart: priorStart,
    metrics: {
      'transactions': _metric(curr.transactions.toDouble(), prev.transactions.toDouble()),
      'totalSpent':   _metric(curr.totalSpent,              prev.totalSpent),
      'taxPaid':      _metric(curr.taxPaid,                 prev.taxPaid),
      'bankFees':     _metric(curr.bankFees,                prev.bankFees),
      'purchases':    _metric(currBank.purchases,           prevBank.purchases),
      'payments':     _metric(currBank.payments,            prevBank.payments),
      'interestPaid': _metric(currBank.interest,            prevBank.interest),
      'feesPaid':     _metric(currBank.fees,                prevBank.fees),
    },
  );
}

/// Fetch bank rows from Supabase (RLS-scoped to the current user).
/// Returns BankData.empty on any failure so callers don't have to
/// branch — the analysis engine handles empty arrays cleanly.
Future<BankData> fetchBankData() async {
  final sb = Supabase.instance.client;
  final user = sb.auth.currentUser;
  if (user == null) return BankData.empty;
  try {
    final results = await Future.wait([
      sb.from('bank_statements').select('*').order('period_end', ascending: false, nullsFirst: false).order('id'),
      sb.from('bank_fees').select('*').order('date', ascending: false, nullsFirst: false).order('id'),
      sb.from('bank_transactions').select('*').order('date', ascending: false, nullsFirst: false).order('id'),
    ]);
    return BankData(
      statements:   List<Map<String, dynamic>>.from(results[0] as List),
      fees:         List<Map<String, dynamic>>.from(results[1] as List),
      transactions: List<Map<String, dynamic>>.from(results[2] as List),
    );
  } catch (_) {
    return BankData.empty;
  }
}
